app/src/main/java/com/github/jorgecastillo/kotlinandroid/functional/Future.kt

```diff
@@ -1,35 +1,68 @@
 package com.github.jorgecastillo.kotlinandroid.functional
 
+import kategory.*
 import kotlinx.coroutines.experimental.CommonPool
 import kotlinx.coroutines.experimental.Deferred
 import kotlinx.coroutines.experimental.android.UI
 import kotlinx.coroutines.experimental.async
 import kotlinx.coroutines.experimental.launch
 
+typealias FutureKind<A> = HK<Future.F, A>
+
+fun <A> FutureKind<A>.ev(): Future<A> =
+        this as Future<A>
+
 /**
  * Basic future implementation to achieve asynchronicity based on Kotlin coroutines.
  */
-class Future<T> {
+class Future<T> : FutureKind<T> {
+
+    class F private constructor()
+
+    private val deferred: Deferred<T>
+
+    private constructor(deferred: Deferred<T>) {
+        this.deferred = deferred
+    }
+
+    constructor(f: () -> T) : this(async(CommonPool) { f() })
 
-  private val deferred: Deferred<T>
+    fun <X> map(f: (T) -> X): Future<X> {
+        return Future(async(CommonPool) { f(deferred.await()) })
+    }
+
+    fun <X> flatMap(f: (T) -> Future<X>): Future<X> {
+        return Future(async(CommonPool) { f(deferred.await()).deferred.await() })
+    }
+
+    fun <B> zip(fb: Future<B>): Future<Tuple2<T, B>> =
+            flatMap { a -> fb.map { b -> Tuple2(a, b) } }
+
+    fun onComplete(f: (T) -> Unit) {
+        launch(UI) {
+            f(deferred.await())
+        }
+    }
 
-  private constructor(deferred: Deferred<T>) {
-    this.deferred = deferred
-  }
+    companion object : Monad<F> {
+        override fun <A, B> flatMap(fa: FutureKind<A>, f: (A) -> FutureKind<B>): Future<B> =
```

Comment:
what's the purpose of adding this comp obejct and the method on it working with `FutureKind` and not `Future`? That's another thing that should clearly not be on this repo but inside `kategory` extras / constructs / whatever separated module. Almost 0 users out there are going to know what's this used for and I wouldn't want them to know, most probably, at least not from the beginning. How hard would be to add a separated module from now on to the lib with these types of constructs like `Future` to avoid too deep FP boilerplate and noise required on client libs / projects?